src/db.rs

```diff
@@ -1021,19 +1021,63 @@ impl Db {
     ///     db.put(b"a", b"a_value").await?;
     ///     db.put(b"b", b"b_value").await?;
     ///
-    ///     let mut iter = db.scan(..).await?;
-    ///     assert_eq!(Some((b"a" as &[u8], b"a_value" as &[u8]).into()) , iter.next().await?);
-    ///     assert_eq!(Some((b"b" as &[u8], b"b_value" as &[u8]).into()) , iter.next().await?);
-    ///     assert_eq!(None , iter.next().await?);
+    ///     let mut iter = db.scan("a".."b").await?;
+    ///     assert_eq!(Some((b"a" as &[u8], b"a_value" as &[u8]).into()), iter.next().await?);
+    ///     assert_eq!(None, iter.next().await?);
     ///     Ok(())
     /// }
     /// ```
-    pub async fn scan<T: RangeBounds<Bytes>>(&self, range: T) -> Result<DbIterator, SlateDBError> {
+    pub async fn scan<K, T>(&self, range: T) -> Result<DbIterator, SlateDBError>
+    where
+        K: AsRef<[u8]>,
+        T: RangeBounds<K>,
+    {
+        let start = range
+            .start_bound()
+            .map(|b| Bytes::copy_from_slice(b.as_ref()));
+        let end = range
+            .end_bound()
+            .map(|b| Bytes::copy_from_slice(b.as_ref()));
+        let range = (start, end);
```

Comment:
Given that the goals of this refactor are (1) API flexibility and (2) avoiding copies, I'm tempted to suggest breaking from `AsRef` only for `scan[_with_options]`, and using `Vec<u8>` instead.

With an `AsRef`, we necessarily have to copy here because even if the `AsRef` owns the value, we don't have a (safe) way of taking the owned value. If instead we passed a `Vec<u8>`, if the caller does want to give up ownership, we avoid an unnecessary copy. If not, they do have to call `.to_vec()` which does make the API a bit clunkier, but at least the copy itself would have happened anyway.

This would also mean that we wouldn't need the `_all` variants, since it's a concrete type.

Reply:
Started to play with this. It looks like `RangeBounds` is requiring a `Sized` type, so `Vec<u8>` won't work here. We'd have to use a `&[u8]`.

Reply:
@hachikuji Would you mind weighing in here? I dug into the `BytesRange` code a bit, and I'm not sure I follow this comment:

```
/// Concrete struct representing a range of Bytes. Gets around much of
/// the cumbersome work associated with the generic trait RangeBounds<Bytes>
```

If we used `RangeBounds<Vec<u8>>` in `scan()`, do we need `BytesRange` anymore, or can we directly use `RangeBounds<Vec<u8>>` everywhere now?

Reply:
@criccomini Yeah, after I refactored things a bit, I realized that `BytesRange` was not doing that much for us anymore. I started a patch to get rid of it, but haven't had time to post it.

Reply:
Awesome, that should simplify things. 😅